### PR TITLE
Multi scheduler support

### LINKED ISSUE #10: Split src and tests
In /bartender the code and the tests are nested. I like it when the tests are organized in the same way as the tests.

So change
* bartender/tests -> tests
* bartender -> src/bartender
* bartender/tests/test_user.py -> tests/web/test_user.py
* bartender/tests/test_bartender.py -> tests/web/test_health.py
* bartender/tests/test_filesystem.py -> tests//filesystem/test_assemble_job.py
* bartender/conftest.py -> tests/conftest.py

### PR DESCRIPTION
**Pull request by sverhoeven**
Uses configuration file to specify applications, schedulers and filesystems.
By default uses first scheduler and first filesystem to upload input, submit, state fetch, download output.

To test the multi scheduler support you have to create a config.yaml file by copying config-example.yaml. The example config file uses a in-memory scheduler on the local file system. To use the Slurm scheduler and sftp filesystem 
1. you have to uncomment the `slurmcontainer` block and 
2. start up a Slurm Docker container as described in the config file.
3. you should be able to submit a job that
    1.  Copies input files to container in `/home/xenon/<job_id>`, which you can verify with `docker exec <container id> ls -l /home/xenon/<job_id>`
    2. Submit job to Slurm, which you can verify by running `sacct` to see the completed job.
    3. After job completion will copy the output files back to where `bartender serve` is running, which you can verify with `ls /tmp/jobs/*`, it should have a stdout.txt file.

Also
* Adds Sphinx doc site (#13 )
* fixes #10

### CODE REVIEW

File: tests/schedulers/test_build.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+        expected = MemoryScheduler(config)
+        assert result == expected
+    finally:
+        await result.close()
+        await expected.close()
+
+
+@pytest.mark.anyio
+async def test_single_localsimplist_slurm_scheduler() -> None:
```
Comment (Peter9192) [inline tests/schedulers/test_build.py:23]: what is localsimplist?

File: config-haddock3.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+destination_picker: bartender.config:pick_first
+applications:
+  haddock3:
+    command: haddock3 $config
+    config: workflow.cfg
```
[the remark this answers is not in the document]
Comment (Peter9192) [reply config-haddock3.yaml:5]: update: if I understand correctly this is a special case, so should we document that?

**Review by Peter9192 — COMMENTED** [review]
Finished going through the code, will continue with some tests

File: config-example.yaml [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# By default the files of jobs are stored in /tmp/jobs
+# job_root_dir: /tmp/jobs
+# By default jobs are submitted to the first destination
+# destination_picker: bartender.picker:pick_first
+# To use a custom picker set `destination_picker` to a `<module>:<function>`
+# The picker should have type bartender.picker.DestinationPicker .
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply config-example.yaml:6]: see #39

File: config-example.yaml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    # remote:
+    #   scheduler:
+    #     type: slurm
+    #     partition: thin
+    #     ssh_config:
+    #       hostname: snellius.surf.nl
+    #   filesystem:
+    #     type: sftp
+    #     ssh_config:
+    #       hostname: snellius.surf.nl
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply config-example.yaml:32]: Yes, no password == public key without passphrase or ssh agent.

File: config-example.yaml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    #       type: slurm
+    #       partition: mypartition
+    #       ssh_config:
+    #           port: 10022
+    #           hostname: localhost
+    #           username: xenon
+    #           password: javagat
+    #   filesystem:
+    #       type: sftp
+    #       ssh_config:
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply config-example.yaml:47]:
```yaml
ssh_configurations:
  snellius:
    hostname: snellius.surf.nl
   ...
destinations:
 bla:
   filesystem:
     ssh_config: 'snellius'
     ...
```

or 

```yaml
destinations:
 bla:
   scheduler:
      type: slurm
      ssh_config:
        hostname: ...
      filesystem:
       type: sftp
       # will get initialized with ssh_config of scheduler
```

[the remark this answers is not in the document]
Comment (sverhoeven) [reply config-example.yaml:47]: https://github.com/i-VRESSE/bartender/issues/40

File: config-example.yaml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    # Name of config file that the application needs
+    config: README.md
+destinations:
+  local:
+    scheduler:
+      type: memory
+      slots: 1
+    filesystem:
+      type: local
+    ## Example of running jobs by current user on snellius.
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply config-example.yaml:22]: #39

File: config-example.yaml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    ## Start a container with `docker run --detach --publish 10022:22 xenonmiddleware/slurm:20`
+    # slurmcontainer:
+    #   scheduler:
+    #       type: slurm
+    #       partition: mypartition
+    #       ssh_config:
+    #           port: 10022
+    #           hostname: localhost
+    #           username: xenon
+    #           password: javagat
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply config-example.yaml:44]: Ok for demo and development, but not for normal use. see https://github.com/i-VRESSE/bartender/issues/41

File: src/bartender/picker.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+DestinationPicker = Callable[[Path, str, "Context"], str]
+
+
+def pick_first(
+    job_dir: Path,
+    application_name: str,
+    context: "Context",
+) -> str:
+    """Always picks first destination to where a job should be submitted to.
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply src/bartender/picker.py:16]: done

File: src/bartender/picker.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        context: "Context",
+    ) -> str:
+        """Always picks the next destination to where a job should be submitted to.
+
+        Takes list of destinations and each time it is called will
+        pick the next destination in the destination list.
+        Going around to start when end is reached.
+
+        :param job_dir: Location where job input files are located.
+        :param application_name: Application name that should be run.
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply src/bartender/picker.py:46]: More advanced picker might want to use those.

File: tests/schedulers/test_build.py (hunk 1 above)
Comment (sverhoeven) [reply tests/schedulers/test_build.py]: removed

File: src/bartender/schedulers/memory.py [hunk 10]
```diff
@@ -6,17 +6,22 @@
...
+    from pydantic.dataclasses import dataclass  # noqa: WPS440
+
 from pydantic.types import PositiveInt
 
 from bartender.db.models.job_model import CompletedStates, State
 from bartender.schedulers.abstract import AbstractScheduler, JobDescription
 
 
-@dataclass
+@pydataclass
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply src/bartender/schedulers/memory.py]: Now prefer BaseModel

File: docs/index.rst [hunk 11]
```diff
@@ -0,0 +1,20 @@
+.. bartender documentation master file, created by
+   sphinx-quickstart on Fri Jan 13 10:01:47 2023.
+   You can adapt this file completely to your liking, but it should at least
+   contain the root `toctree` directive.
+
+Welcome to bartender's documentation!
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply docs/index.rst:6]: Is part of #13

File: docs/index.rst [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+=====================================
+
+.. toctree::
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply docs/index.rst:9]: Added to #13

File: src/bartender/config.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
+"""
+Parses config file into a list of applications, schedulers and filesystems.
+
+Example config:
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply src/bartender/config.py:4]: See #39

File: src/bartender/config.py [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+    with value of ApplicatonConfiguration.config.
+
+    .. code-block:: yaml
+
+        command: wc $config
+        config: README.md
+    """
+
+    command: str
+    config: str
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply src/bartender/config.py:83]: Yes

File: src/bartender/config.py [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+    # TODO make config optional,
+    # as some commands don't need a config file name as argument
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply src/bartender/config.py:85]: See #43 #42

File: src/bartender/filesystems/local.py [hunk 16]
```diff
@@ -1,9 +1,18 @@
...
 
 from bartender.filesystems.abstract import AbstractFileSystem
 from bartender.schedulers.abstract import JobDescription
 
 
+@dataclass
+class LocalFileSystemConfig:
+    """Configuration for local file system."""
+
+    type: Literal["local"] = "local"
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply src/bartender/filesystems/local.py:13]: Helpful for doing `LocalFileSystemConfig()` or constructed by destination class (see https://docs.pydantic.dev/usage/types/#discriminated-unions-aka-tagged-unions).

File: src/bartender/config.py [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+    If config is empty will create a single slot in memory scheduler
+    with a local file system.
+
+    """
+
+    applications: dict[str, ApplicatonConfiguration]
+    destinations: dict[str, DestinationConfig] = Field(
+        default_factory=default_destinations,
+    )
+    job_root_dir: DirectoryPath = Path(gettempdir()) / "jobs"
```
Comment (Peter9192) [inline src/bartender/config.py:114]: Doesn't check if exists when using default path

**Review by Peter9192 — APPROVED** [review]
Tests mostly work fine, only check that job root dir exists doens't work for default

File: src/bartender/config.py (hunk 17 above)
Comment (Peter9192) [reply src/bartender/config.py:114]: This works:

```py
from pydantic import BaseModel, DirectoryPath


class TestClass(BaseModel):
    root_dir: DirectoryPath = '/does/not/exist'

    class Config:
        validate_all = True


cfg = TestClass()
print(repr(cfg.root_dir))
```

**Pull request merged**